src/components/code/Code.tsx

```diff
@@ -1,148 +1,139 @@
-import React, {
-  forwardRef,
-  ForwardedRef,
-  useRef,
-  useState,
-  useEffect,
-} from "react"
+import React, { forwardRef, useState } from "react"
 import cn from "classnames"
 
 import { Button } from "../button/Button"
+import { Icon } from "../icon/Icon"
 
-const COPIED_STATUS_TIMEOUT = 3000
+const DEFAULT_SECRET_TEXT = "************"
 
 enum CopiedStatus {
   Success = "SUCCESS",
   Failed = "FAILED",
 }
 
-interface BaseProps {
-  /** Text to be styled as code like representation */
-  children: string
+interface CommonProps {
+  /** Text value to be styled as code like representation */
+  value: string
+  /** prefix?: string (Optional string to render with a non inline Code block) */
+  prefix?: string
   /** Additional classes that will be spread over the icon. Avoid changing the Icon visually. */
   className?: string
 }
 
 /**
- * Discriminated Union to accomodate the following conditional prop logic
- * isInline?: boolean (Optional flag to render inline Code styles)
- * prefix?: string (Optional string to render with a non inline Code block)
- * ref?: A React ref to attach to the rendered Code block's copy button
- * getSecret?: A function that can be used to dynamcally fetch a value
+ * Narrowing conditions, highlighting possible paths of execution
+ * isSecret?: Controls if the Code component holds sensitive data that is not to be revealed immediately
+ * secretText?: Optional masked value for sensitive data.
+ * canReveal?: Stipulates if the user has permission to reveal sensitive data or not
  */
 type ConditionalProps =
-  | { isInline?: true; prefix?: never; ref?: never; getSecret?: never }
   | {
-      isInline?: false
-      prefix?: string
-      ref?: ForwardedRef<HTMLButtonElement>
-      getSecret?: () => string
+      isSecret?: false
+      secretText?: never
+      canReveal?: never
+    }
+  | {
+      isSecret: true
+      secretText?: string
+      canReveal?: boolean
     }
 
-export type CodeProps = BaseProps & ConditionalProps
+export type CodeProps = CommonProps & ConditionalProps
 
 export const Code = forwardRef<HTMLButtonElement, CodeProps>(
   (
-    { isInline = false, prefix, getSecret, children, className }: CodeProps,
-    forwardedRef
+    {
+      value,
+      prefix,
+      className,
+      isSecret,
+      secretText = DEFAULT_SECRET_TEXT,
+      canReveal,
+    }: CodeProps,
+    ButtonRef
   ) => {
-    const _ref = useRef<HTMLButtonElement>(null)
-    const ref = forwardedRef || _ref
+    const [isValueVisible, setIsValueVisible] = useState(
+      isSecret ? false : true
+    )
     const [copied, setCopied] = useState<CopiedStatus | null>(null)
-    const [secretValue, setSecretValue] = useState("")
-    const isSecretVisible = secretValue === ""
+    const canCopyValue = !isSecret || (isSecret && canReveal) ? true : false
```

Comment:
I think you should be able to drop the ? true : false
```suggestion
    const canCopyValue = !isSecret || (isSecret && canReveal)
```